fix Liquibase classloader loading in multiple classloaders scenario

sonersivri opened the pull request:
## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes expected existing functionality)
 
## Description

Liquibase has problem with multiple classloader Cannot find default log service error

## Things to be aware of


## Things to worry about

## Additional Context

Fixes #3819

lorenzopiovesanzextras commented on liquibase-core/src/main/java/liquibase/servicelocator/StandardServiceLocator.java:
```diff
@@ -16,21 +16,40 @@ public int getPriority() {
     @Override
     public <T> List<T> findInstances(Class<T> interfaceType) throws ServiceNotFoundException {
         List<T> allInstances = new ArrayList<>();
-
```
@sonersivri , I think both the above List and below Set can be both final

nvoxland commented:
**Initial/Pre-Review Thoughts**

We don't have example/test code using multiple classloaders to be able to test it out well. But if it impacts the "normal" single-classloader setup that would get caught by the existing tests. And I'm not sure it's worth blocking this fix to build more robust tests against it.

The original issue has a good example of the project setup that was hitting this problem.

Questions I have:
- None

Potential risks:
- None

What could make the full review difficult:
(List what could trip us up and burn time during the review process)
- Nothing

sonersivri commented:
> **Initial/Pre-Review Thoughts**
> 
> We don't have example/test code using multiple classloaders to be able to test it out well. But if it impacts the "normal" single-classloader setup that would get caught by the existing tests. And I'm not sure it's worth blocking this fix to build more robust tests against it.
> 
> The original issue has a good example of the project setup that was hitting this problem.
> 
> Questions I have:
> 
> * None
> 
> Potential risks:
> 
> * None
> 
> What could make the full review difficult: (List what could trip us up and burn time during the review process)
> 
> * Nothing

This method aim is find implemented class I only added look at also given Class classLoader if classLoader is not same as method parameters classLoader.
This should not be problematic.
In our cases we create big jar with dependencies and using it in plugin system of carbonio.

filipelautert commented:
Hello 
I tried to reproduce it with a simple JakartaEE project that I created at https://github.com/filipelautert/liquibase-jakarta-demo . The "main" class is https://github.com/filipelautert/liquibase-jakarta-demo/blob/main/src/main/java/com/example/liqiubasejakartademo/liquibase/LiquibaseStartupUpdate.java .

I tried in 2 ways: 
* Starting a database test container and creating the jdbc connection
* Getting a connection from a server datasource
 *Tried with Liquibase 4.19 and 4.22 

It worked fine in the 2 cases. Do you have any idea of what I could change to reproduce it? Thanks!

kevin-atx commented:
@sonersivri - it looks like @filipelautert was unable to reproduce your issue. Would you be able to provide more detailed steps to reproduce?

kevin-atx commented:
@tati-qalified - could you try to reproduce this issue?

sonersivri commented:
> @sonersivri - it looks like @filipelautert was unable to reproduce your issue. Would you be able to provide more detailed steps to reproduce?

@kevin-atx @filipelautert you can reproduce using https://github.com/sonersivri/liquibase-bug-3820

filipelautert reviewed (approved):
hey @sonersivri , thanks for the example project. I can't see how we could reproduce it using junit as you're doing some serious manual classloading there. 
Anyway, I updated you PR to our current base (just I touched this class so it was simple to fix the merge conflicts) and added some comments trying to explain why you added the second call to findInstances. Could you review them?

And thanks again for the PR!

sonersivri commented:
> hey @sonersivri , thanks for the example project. I can't see how we could reproduce it using junit as you're doing some serious manual classloading there. Anyway, I updated you PR to our current base (just I touched this class so it was simple to fix the merge conflicts) and added some comments trying to explain why you added the second call to findInstances. Could you review them?
> 
> And thanks again for the PR!

@filipelautert Thanks for review and checks. I checked the PR it looks ok. 
I also test new code in my project it works fine.

sonersivri commented:
Guys is there any progress about our company planned to change the using liquibase

sonersivri commented:
My solution is not solve our problem I am closing the PR
